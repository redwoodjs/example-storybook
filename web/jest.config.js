// More info at https://redwoodjs.com/docs/project-configuration-dev-test-build

const config = {
  rootDir: '../',
  preset: '@redwoodjs/testing/config/jest/web',
  verbose: true,
  moduleNameMapper: {
    // mock redwood magic webpack aliases
    // see: https://github.com/redwoodjs/redwood/blob/main/packages/testing/config/storybook/preview.js
    '~__REDWOOD__USER_WEB_DEFAULT_CSS': 'identity-obj-proxy',
    '~__REDWOOD__USER_STORYBOOK_PREVIEW_CONFIG':
      '<rootDir>/web/config/storybook.preview.js',
  },
  transform: {
    '^.+\\.mdx?$': '@storybook/addon-docs/jest-transform-mdx', // https://github.com/storybookjs/storybook/issues/11376#issuecomment-757417250
  },
  setupFilesAfterEnv: [`<rootDir>/web/test/mocks.js`],
}

module.exports = config
