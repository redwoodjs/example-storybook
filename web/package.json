{
  "name": "web",
  "version": "0.0.0",
  "private": true,
  "browserslist": {
    "development": [
      "last 1 version"
    ],
    "production": [
      "defaults",
      "not IE 11",
      "not IE_Mob 11"
    ]
  },
  "dependencies": {
    "@redwoodjs/forms": "0.49.0",
    "@redwoodjs/router": "0.49.0",
    "@redwoodjs/web": "0.49.0",
    "prop-types": "15.8.1",
    "react": "17.0.2",
    "react-dom": "17.0.2"
  }
}
