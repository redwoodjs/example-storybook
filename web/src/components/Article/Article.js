import { Link, routes } from '@redwoodjs/router'

const Article = ({ article }) => {
  return (
    <article>
      <header>
        <h2 className="text-xl text-blue-700 font-semibold">
          <Link to={routes.article({ id: article.id })}>{article.title}</Link>
        </h2>
      </header>
      <div className="mt-2 text-gray-900 font-light">{article.body}</div>
      <div>Posted at: {article.createdAt}</div>
    </article>
  )
}

export default Article
