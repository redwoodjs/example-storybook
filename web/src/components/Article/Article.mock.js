// Define your own mock data here:
export const standard = () => ({
  article: {
    id: 42,
    title: 'Article Title',
    body: 'Article Body',
    createdAt: '2020-01-01T00:00:00.000Z',
  },
})
