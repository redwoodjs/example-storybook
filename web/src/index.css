/**
 * START --- SETUP TAILWINDCSS EDIT
 *
 * `yarn rw setup ui tailwindcss` placed these imports here
 * to inject Tailwind's styles into your CSS.
 * For more information, see: https://tailwindcss.com/docs/installation#include-tailwind-in-your-css
 */
@import 'tailwindcss/base';
@import 'tailwindcss/components';
@import 'tailwindcss/utilities';
/**
 * END --- SETUP TAILWINDCSS EDIT
 */
body {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
ul {
  list-style-type: none;
  margin: 1rem 0;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 1rem 0 0 ;
}
h1 > a {
  text-decoration: none;
  color: black;
}
button, input, label, textarea {
  display: block;
  outline: none;
}
label {
  margin-top: 1rem;
}
.error {
  color: red;
}
input.error, textarea.error {
  border: 1px solid red;
}
.form-error {
  color: red;
  background-color: lavenderblush;
  padding: 1rem;
  display: inline-block;
}
.form-error ul {
  list-style-type: disc;
  margin: 1rem;
  padding: 1rem;
}
.form-error li {
  display: list-item;
}
.flex-between {
  display: flex;
  justify-content: space-between;
}
.flex-between button {
  display: inline;
}
