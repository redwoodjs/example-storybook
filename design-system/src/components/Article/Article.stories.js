import { mockGraphQLMutation } from '@redwoodjs/testing/web'

import Article from './Article'
import { standard } from './Article.mock'

/** @type { import("@storybook/react").Meta } */
export default {
  title: 'Components/Article',
}

export const full = (args) => {
  mockGraphQLMutation('CreateCommentMutation', (variables, { ctx }) => {
    const id = parseInt(Math.random() * 1000)
    ctx.delay(1000)

    return {
      createComment: {
        id,
        name: variables.input.name,
        body: variables.input.body,
        createdAt: new Date().toISOString(),
      },
    }
  })
  return <Article {...standard()} {...args} />
}

export const summary = ({ summary = true }) => {
  return <Article article={standard().article} summary={summary} />
}
