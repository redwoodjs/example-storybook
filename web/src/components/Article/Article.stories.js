import Article from './Article'
import { standard } from './Article.mock'

/** @type { import("@storybook/react").Meta } */
export default {
  title: 'Components/Article',
}

export const full = (args) => {
  return <Article {...standard()} {...args} />
}

export const summary = ({ summary = true }) => {
  return <Article article={standard().article} summary={summary} />
}
