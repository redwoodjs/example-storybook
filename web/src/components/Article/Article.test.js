import { screen, render, waitFor } from '@redwoodjs/testing/web'
import { standard } from './Article.mock'

import Article from './Article'

//   Improve this test with help from the Redwood Testing Doc:
//    https://redwoodjs.com/docs/testing#testing-components

describe('Article', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<Article article={standard().article} />)
    }).not.toThrow()
  })

  it('fragment matches snapshot', () => {
    const { asFragment } = render(<Article article={standard().article} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('fragment when summary prop set to true matches snapshot', () => {
    const { asFragment } = render(
      <Article summary={true} article={standard().article} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('fragment when summary prop set to false matches snapshot', () => {
    const { asFragment } = render(
      <Article summary={false} article={standard().article} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('fragment when summary prop set to false and async/await matches snapshot', async () => {
    const { asFragment } = render(
      <Article summary={false} article={standard().article} />
    )
    await waitFor(() => {
      expect(
        screen.getByText('Cowabunga dudes. This is the first comment!')
      ).toBeInTheDocument()
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
