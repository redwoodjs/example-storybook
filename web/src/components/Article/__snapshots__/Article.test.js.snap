// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article fragment matches snapshot 1`] = `
<DocumentFragment>
  <article>
    <header>
      <h2
        class="text-xl text-blue-700 font-semibold"
      >
        <a
          href="/article/1"
        >
          First Post
        </a>
      </h2>
    </header>
    <div
      class="mt-2 text-gray-900 font-light"
    >
      Neutra tacos hot chicken prism raw denim, put a bird on it enamel pin post-ironic vape cred DIY. Street art next level umami squid. Hammock hexagon glossier 8-bit banjo. Neutra la croix mixtape echo park four loko semiotics kitsch forage chambray. Semiotics salvia selfies jianbing hella shaman. Letterpress helvetica vaporware cronut, shaman butcher YOLO poke fixie hoodie gentrify woke heirloom.
    </div>
    <div>
      Loading...
    </div>
    <div>
      Posted at: 2022-03-30T00:00:00.000Z
    </div>
  </article>
</DocumentFragment>
`;

exports[`Article fragment when summary prop set to false and async/await matches snapshot 1`] = `
<DocumentFragment>
  <article>
    <header>
      <h2
        class="text-xl text-blue-700 font-semibold"
      >
        <a
          href="/article/1"
        >
          First Post
        </a>
      </h2>
    </header>
    <div
      class="mt-2 text-gray-900 font-light"
    >
      Neutra tacos hot chicken prism raw denim, put a bird on it enamel pin post-ironic vape cred DIY. Street art next level umami squid. Hammock hexagon glossier 8-bit banjo. Neutra la croix mixtape echo park four loko semiotics kitsch forage chambray. Semiotics salvia selfies jianbing hella shaman. Letterpress helvetica vaporware cronut, shaman butcher YOLO poke fixie hoodie gentrify woke heirloom.
    </div>
    <div
      class="space-y-8"
    >
      <div
        class="bg-gray-200 p-8 rounded-lg"
      >
        <header
          class="flex justify-between"
        >
          <h2
            class="font-semibold text-gray-700"
          >
            April O'Neil
          </h2>
          <time
            class="text-xs text-gray-500"
            datetime="2022-04-01T12:34:56Z"
          >
            1 April 2022
          </time>
        </header>
        <p
          class="text-sm mt-2"
        >
          Cowabunga dudes. This is the first comment!
        </p>
      </div>
      <div
        class="bg-gray-200 p-8 rounded-lg"
      >
        <header
          class="flex justify-between"
        >
          <h2
            class="font-semibold text-gray-700"
          >
            Michelangelo
          </h2>
          <time
            class="text-xs text-gray-500"
            datetime="2020-02-03T23:00:00Z"
          >
            3 February 2020
          </time>
        </header>
        <p
          class="text-sm mt-2"
        >
          It's üçï time!
        </p>
      </div>
    </div>
    <div>
      Posted at: 2022-03-30T00:00:00.000Z
    </div>
  </article>
</DocumentFragment>
`;

exports[`Article fragment when summary prop set to false matches snapshot 1`] = `
<DocumentFragment>
  <article>
    <header>
      <h2
        class="text-xl text-blue-700 font-semibold"
      >
        <a
          href="/article/1"
        >
          First Post
        </a>
      </h2>
    </header>
    <div
      class="mt-2 text-gray-900 font-light"
    >
      Neutra tacos hot chicken prism raw denim, put a bird on it enamel pin post-ironic vape cred DIY. Street art next level umami squid. Hammock hexagon glossier 8-bit banjo. Neutra la croix mixtape echo park four loko semiotics kitsch forage chambray. Semiotics salvia selfies jianbing hella shaman. Letterpress helvetica vaporware cronut, shaman butcher YOLO poke fixie hoodie gentrify woke heirloom.
    </div>
    <div>
      Loading...
    </div>
    <div>
      Posted at: 2022-03-30T00:00:00.000Z
    </div>
  </article>
</DocumentFragment>
`;

exports[`Article fragment when summary prop set to true matches snapshot 1`] = `
<DocumentFragment>
  <article>
    <header>
      <h2
        class="text-xl text-blue-700 font-semibold"
      >
        <a
          href="/article/1"
        >
          First Post
        </a>
      </h2>
    </header>
    <div
      class="mt-2 text-gray-900 font-light"
    >
      Neutra tacos hot chicken prism raw denim, put a bird on it enamel pin post-ironic vape cred DIY. Str...
    </div>
    <div>
      Posted at: 2022-03-30T00:00:00.000Z
    </div>
  </article>
</DocumentFragment>
`;
